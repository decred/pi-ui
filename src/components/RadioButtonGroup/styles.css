.radioButton {
  display: inline-block;
  position: relative;
  padding: 0 0.6rem;
  margin: 1rem 1rem 0 0;
  user-select: none;
}

.radioButton input[type="radio"] {
  opacity: 0;
  position: fixed;
}

.radioButton .circle {
  display: inline-block;
  position: relative;
  top: 0.5rem;
  margin: 0 0.5rem 0 0;
  width: 2rem;
  height: 2rem;
  border-radius: 1.1rem;
  border: 0.2rem solid var(--color-primary-dark);
  background-color: transparent;
}

.radioButton input[type="radio"]:focus + label > .circle {
  box-shadow: 0 0 0.3rem 0.1rem var(--color-primary-dark);
}

.radioButton input[type="radio"]:disabled {
  pointer-events: none;
}

.radioButton .dot {
  border-radius: 1.1rem;
  width: 1.2rem;
  height: 1.2rem;
  position: absolute;
  top: 0.9rem;
  left: 1rem;
  display: block;
  background: var(--color-primary-dark);
}

.radioButton.disabled .dot {
  opacity: 0.6;
}

.radioGroupList {
  list-style-type: none;
  display: flex;
}

.vertical {
  display: block;
}

.radioGroupLabel {
  font-size: var(--font-size-small);
  color: var(--color-gray);
}
