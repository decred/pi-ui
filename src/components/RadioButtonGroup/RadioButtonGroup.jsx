import React from "react";
import PropTypes from "prop-types";
import { classNames } from "../../utils";
import styles from "./styles.css";

export const RadioButton = ({
  name,
  onChange,
  onBlur,
  id,
  checked,
  label,
  className,
  ...props
}) => {
  return (
    <div className={classNames(styles.radioButton, className)}>
      <input
        name={name}
        id={id}
        type="radio"
        checked={checked}
        onChange={onChange}
        onBlur={onBlur}
        className={classNames(styles.radioButton, className)}
        {...props}
      />
      <label htmlFor={id}>{label}</label>
    </div>
  );
};

export const RadioButtonGroup = ({
  options,
  value,
  onChange,
  label,
  name,
  className,
  vertical,
  optionsListClassName,
  optionClassName
}) => {
  return (
    <div className={classNames(styles.radioGroup, className)}>
      <label className={styles.radioGroupLabel}>{label}</label>
      <ul
        className={classNames(
          styles.radioGroupList,
          vertical && styles.vertical,
          optionsListClassName
        )}>
        {options.map((option) => (
          <li>
            {" "}
            <RadioButton
              onChange={() => onChange(option)}
              className={optionClassName}
              checked={option.value === value}
              label={option.label}
              name={name}
            />
          </li>
        ))}
      </ul>
    </div>
  );
};

RadioButton.propTypes = {
  name: PropTypes.string,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  id: PropTypes.string,
  checked: PropTypes.bool,
  label: PropTypes.string,
  className: PropTypes.string
};

RadioButtonGroup.propTypes = {
  options: PropTypes.array.isRequired,
  value: PropTypes.any,
  onChange: PropTypes.func.isRequired,
  label: PropTypes.string,
  name: PropTypes.string,
  className: PropTypes.string,
  vertical: PropTypes.bool,
  optionsListClassName: PropTypes.string,
  optionClassName: PropTypes.string
};
