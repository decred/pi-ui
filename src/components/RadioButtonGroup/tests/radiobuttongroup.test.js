import React from "react";
import { create } from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import { RadioButtonGroup } from "../RadioButtonGroup";

const options = [
  { label: "foo", value: 1 },
  { label: "bar", value: 2 }
];
describe("RadioButtonGroup component", () => {
  test("Matches snapshot", () => {
    const radiogroup = create(
      <RadioButtonGroup
        label="Test"
        options={options}
        value={1}
        onChange={jest.fn()}
      />
    );
    expect(radiogroup.toJSON()).toMatchSnapshot();
  });

  test("RadioButton click triggers change function and updates value", () => {
    let value = 1;
    const mockHandleChange = jest.fn(() => {
      value = 2;
    });
    const { getByLabelText, queryByLabelText } = render(
      <RadioButtonGroup
        label="Test"
        options={options}
        value={value}
        onChange={mockHandleChange}
      />
    );
    expect(queryByLabelText(/foo/i)).toBeTruthy();
    expect(queryByLabelText(/bar/i)).toBeTruthy();
    fireEvent.click(getByLabelText(/bar/i));
    expect(mockHandleChange).toBeCalled();
    expect(value).toBe(2);
  });
});
