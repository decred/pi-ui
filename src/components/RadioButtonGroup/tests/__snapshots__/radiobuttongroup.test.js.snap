// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroup component Matches snapshot 1`] = `
<div
  className="radioGroup"
>
  <label
    className="radioGroupLabel"
  >
    Test
  </label>
  <ul
    className="radioGroupList"
  >
    <li>
       
      <div
        className="radioButton"
      >
        <input
          checked={true}
          onChange={[Function]}
          type="radio"
        />
        <label
          onClick={[Function]}
        >
          <span
            className="circle"
          />
          foo
          <span
            className="dot"
            style={
              Object {
                "transform": "scale(1)",
              }
            }
          />
        </label>
      </div>
    </li>
    <li>
       
      <div
        className="radioButton"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="radio"
        />
        <label
          onClick={[Function]}
        >
          <span
            className="circle"
          />
          bar
          <span
            className="dot"
            style={
              Object {
                "transform": "scale(0)",
              }
            }
          />
        </label>
      </div>
    </li>
  </ul>
</div>
`;
